### PR TITLE
Compute Shannon entropy of files, for bytes, bi-grams and tri-grams.

### PR DESCRIPTION
**Pull request by nishihatapalmer**
Here's a small PR that computes the Shannon entropy of the bytes, bi-grams and tri-grams for each file that's read, along with the counts of each.

I'm interested not only in the alphabet used by data, but the entropy the data has. 

Feel free to ignore if you don't find this useful.

### CODE REVIEW

File: source/smart.c [hunk 1]
```diff
@@ -123,17 +123,71 @@ int getText(unsigned char *T, char *path, int FREQ[SIGMA], int TSIZE) {
...
-		if(maxcode<T[j]) maxcode=T[j];
-	}
-	printf("\tAlphabet of %d characters.\n", nalpha);
-	printf("\tGreater chararacter has code %d.\n", maxcode);
+    // compute the frequency of characters and the dimension of the alphabet
+    int nalpha = 0;
+    int nbigram = 0;
+    int ntrigram = 0;
+    int maxcode= 0;
+    for(j=0; j<SIGMA; j++) FREQ[j]=0;
```
Comment (Helias) [inline source/smart.c]: would be better to wrap this into a function like
```C++
void initArr(int *arr, int length) {
  for(int i = 0; i < length; i++) {
    arr[i] = 0;
  }
}
```

and reuse it any time you have to initialize an arr? an inline for can be confusing sometimes

File: source/smart.c [hunk 2]
```diff
@@ -123,17 +123,71 @@ int getText(unsigned char *T, char *path, int FREQ[SIGMA], int TSIZE) {
...
+        }
+        byteValue = T[j];
+        if(FREQ[byteValue]==0) nalpha++;
+        FREQ[byteValue]++;
+        if(maxcode<byteValue) maxcode=byteValue;
+    }
+    // compute the shannon entropy of the text on the bytes:
+    double shannonEntropy = 0.0;
+    double log2 = log(2);
+    for(j=0;j<SIGMA;j++) {
```
Comment (Helias) [inline source/smart.c]: I would always write more space to improve the readability
```suggestion
    for (j = 0; j < SIGMA; j++) {

```

**Comment by nishihatapalmer** [comment]
I've done as you suggested, and extracted the array initialization into a method, and spaced the elements for readability.

I should say - I much prefer elements to be spaced in code; I was trying to fit in with the "house style" of the adjacent code - which doesn't have a lot of spaces. Happy to write in a more readable way.

**Comment by nishihatapalmer** [comment]
Also refactored the computation of the Shannon entropy.  Looks a bit cleaner now.

**Pull request closed without merging**